### ISSUE TITLE
Some problems with Code highlight & Searching speed

### ISSUE THREAD

**Issue by imbadark** [probberechts/hexo-theme-cactus#141]
Using this theme for a long time and I often find code highlighting somewhat problematic... After I cannot get the answer from the issue , I open a issue for those who have same problem.
There are two main problems:
1. In theory, the `cactus` theme should use the official `highlight.js` for code highlighting (I guess)?  however, it will be found that some highlighting languages ​​supported in `highlight.js` are not able to display highlights on the page at all. Using `F12` to view html found that it is not treated as a `sql` code snippet. SO HOW COULD I FIX IT?
2. There are some languages ​​that can be identified as being highlighted like `bash`, and you will find that if there are multiple `bash code` snippets on the same page, some parts cannot be highlighted normally.
I believe it's very likely that some characters affect the highlighting logic. But I don't know how to determine what it is, I found the `highlight.js`'s path on `node_modules/highlight.js/lib/languages`  and check the js-files...  
---
I think some of the css styles that come with the theme won't affect whether the code in some languages ​​can be highlighted or parsed. So can you help me to figure out how to locate these problems?

**Comment by imbadark** [comment]
Add  some questions :
1. One is that when there are more blog posts, the generated search.xml file may have a few megabytes, and each time the search page is accessed, the xml file will be loaded, causing the search function to be particularly slow. (The browser doesn't seem to cache `search,xml`..)
2. The other is that I found that I can't comment out the `open_graph` and google analytics in the theme configuration.( `config.yml` ) ,if I comment on them, `hexo g ` will generate a lot of errors. If I just don't care about them, it seems that after the page is generated, it will load some strange google's urls, and some strange font files, which greatly affect the page loading speed, so I would like to know how to get rid of them.
The urls like : 
- `https://fonts.gstatic.com/s/notosanssc/v2/k3kXo84MPvpLmixcA63oeALhLOCT-xWNm8Hqd37g1OkDRZe7lR4sg1IzSy-MNbE9VH8V.116.woff2`  (a string of urls from `107~120`...)

**Comment by probberechts** [comment]
I don't experience any problems with code highlighting. For example for SQL and bash, i get these results:

```
{% codeblock lang:sql %}
SELECT EMP_ID, LAST_NAME FROM EMPLOYEE
WHERE CITY = 'Seattle' ORDER BY EMP_ID;
{% endcodeblock %}

{% codeblock lang:bash %}
#!/bin/bash          
echo "Hello World"  
{% endcodeblock %}
```
-->
<img width="751" alt="screenshot 2018-12-01 at 11 37 17" src="https://user-images.githubusercontent.com/2175271/49327138-806a1680-f55d-11e8-895c-21607234b95b.png">
 
I'm afraid that I can't help you if I don't know exactly what causes your problem.

**Comment by probberechts** [comment]
The `search.xml` file is loaded via a `jQuery.ajax()` call, [which enables caching by default](http://api.jquery.com/jQuery.ajax/#options). So it's actually your browser that decides whether the `search.xml` file is served from cache or not. I don't see a reason to try to override this default behavior. Also, the `search.xml` file is only a couple of kBytes for a normal blog. I wonder how huge your blog is that you get a file that is a few MB?

**Comment by probberechts** [comment]
The open graph tags are just meta information, they don't load anything. If you don't care about them, then just leave them empty in your config file (which is the default):
```
open_graph:
  fb_app_id:
  fb_admins:
  twitter_id:
  google_plus:
```
If you don't want them, you can remove them from the `layout/_partial/head.ejs` file. But they are useful (e.g., when somebody would like to share some of your posts on social media), so I would advise to keep them.

Google analytics is disabled by default. You can do it explicitly with:
```
google_analytics:
  enabled: false
```

This them doesn't load any google fonts or other google URLs (if you disable Google analytics). A guess it's actually one of your browser plugins that loads these.

**Comment by imbadark** [comment]
> Thank you very much for your detailed reply, I appreciate it.
These days are very busy, so please allow me to briefly answer your feedback.  
1. I found some reasons for the highlighting problem, such as the `SQL` code block, because the old version of `highlight.js` does not seem to support `#` as a comment, so if I use it at the beginning of the code block or in the mediation, then the highlight will  all failed...  I think that the sudden failure of bash and other languages ​​is likely to be a similar reason. I think of two ways to solve this problem: 
    - The first, update the version of `highlight.js` to the latest V9.13 (it comes with V9.11), but I found that npm update does not work, then I tried to update npm Version and package, now have some questions, and I will give detailed feedback after I get it.
     -  Another way is to manually modify the `sql.js` or `bash.js` file to add the keywords that the user wants to support/customize. But I don't know if this method is reasonable, you need to reinstall it, this may be Refer to the issue of `highlight.js`, free to try and give feedback..
2. My blog is probably about a hundred (<150) articles, although some of them are encrypted with the `hexo-blog-encrypt` plugin (maybe it have anything  related to this?) The current `search.xml` file is almost 3MB in size, not very good on the network. When the search function is almost collapsed, this will continue to check the root cause. Thank you for your reminder.
3. Google analysis and graph problems have been solved. (Or ignore it.. )

Finally, remind everyone, do not use `npu -u && npm install` to upgrade the various npm modules of `hexo`, it is easy to have a series of strange parsing problems. Must also be a full backup before the upgrade..

**Comment by imbadark** [comment]
Update:
1. To those who wish to modify or update the highlighted logic/version, the easiest way I found was to download the official xx9.13.1.tgz package, then manually extract it, overwrite the previous file under hexo/node_modules/highlight.js, at last `npm install` to complete it. (You will find that for some reason, direct `npm update highlight.js@9.13.1` does not work)
2. After the update, please first do `hexo g `to ensure that there is no problem and backup files, then start to modify the corresponding `language.js` file according to your needs, (such as `bash.js`), you can easily imitate the existing way to DIY what you want to highlight, but please **be aware** that there are some changes that may cause the `hexo g `to be stuck. For example, `json.js `does not support comments originally, but if you want to manually make `#` and `//` recognized as highlights of the comment, and u copy the function called directly from the `bash.js` file, which may cause the hexo g to be unavailable. (and hard to find...)for more details, u should read the `highlight.js`'s docs carefully.
3. I carefully checked the `search.xml` file my blog generated. I found that the encryption plugin does not expand the volume, conversely it makes it a lot smaller(Because there is no expansion)... The main reason why the file is so large is that it generates a lot of html code (eg. `<span> <br> <pre>`, instead of just loading the original markdown file content. (That would be very small,) I don't know if this is normal? Or can I avoid or clean them by setting the parameters of the `hexo-search` plugin? 
Looking forward to your reply, thank you again. @probberechts

**Comment by probberechts** [comment]
Ok. Thanks.

Regarding 3, you can remove the body of your blog posts from the `search.xml` file with the config below. That will make the `search.xml` file a lot smaller, but you won't be able to search through the body of your posts anymore.
```yml
search:
  path: search.xml
  content: false
```

I agree that all that html code does not belong in the `search.xml` file, but that's handled by the [hexo-generator-search plugin](https://github.com/wzpan/hexo-generator-search). If you would like to see this done differently, you should create an issue over there.

**Comment by imbadark** [comment]
> 好。谢谢。
> 
> 关于3，您可以`search.xml`使用下面的配置从文件中删除博客文章的正文。这将使`search.xml`文件变小，但您将无法再搜索帖子的正文。
> 
> ```yaml
> search：
>    path：search.xml 
>   内容：false
> ```
> 
> 我同意所有html代码都不属于该`search.xml`文件，但这是由[hexo-generator-search插件处理的](https://github.com/wzpan/hexo-generator-search)。如果您希望以不同的方式完成此操作，则应在此处创建问题。

@probberechts The way you set the content to false above will cause the search title to not work(any input in search bar will return `No results found.` ). I don't know why. The console didn't report an error. How can I confirm where the problem is? (When I set it to true, it's OK. )

And I found that the author of the original plugin has read the local source file content directly by generating the search.json file. [e.g. search.json](https://github.com/wzpan/hexo-generator-search/blob/master/demo_output/search.json) 
But the key problem is that I found that the current `search.js` in the theme only supports `search.xml`, which seems to be not updated. Since json's files can narrow down the search index file by more than 10 times, I hope u can consider updating it to support it. THX.

BTW:
